pkg/cluster/config/encoding/scheme.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package encoding
+
+import (
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	"sigs.k8s.io/kind/pkg/cluster/config"
+	"sigs.k8s.io/kind/pkg/cluster/config/v1alpha1"
+)
+
+// Scheme is the runtime.Scheme to which all `kind` config API versions and types are registered.
+var Scheme = runtime.NewScheme()
+
+// Codecs provides access to encoding and decoding for the scheme.
+var Codecs = serializer.NewCodecFactory(Scheme)
+
+func init() {
+	AddToScheme(Scheme)
+}
+
+// AddToScheme builds the scheme using all known `kind` API versions.
+func AddToScheme(scheme *runtime.Scheme) {
+	utilruntime.Must(config.AddToScheme(scheme))
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1alpha1.SchemeGroupVersion))
+}
+
+// newDefaultedConfig creates a new, defaulted `kind` Config
+// Defaulting uses Scheme registered defaulting functions
+func newDefaultedConfig() *config.Config {
+	var config = &config.Config{}
+
+	// apply defaults
+	Scheme.Default(config)
+
+	return config
+}
+
+// unmarshalConfig attempt to decode data into a `kind` Config; data can be
+// one of the different API versions defined in the Scheme.
+func unmarshalConfig(data []byte) (*config.Config, error) {
+	var config = &config.Config{}
+
+	// decode data into a config object
+	_, _, err := Codecs.UniversalDecoder().Decode(data, nil, config)
+	if err != nil {
+		return nil, errors.Wrap(err, "decoding failure")
+	}
+
+	// apply defaults
+	Scheme.Default(config)
```

Comment:
So in other Kubernetes projects, typically the application itself works with a particular API version.

For example, kube-controller-manager works with the versioned representation of types.

From my understanding, the universal decoder should be able to just directly decode into the versioned representation.

Reply:
this is how it's done in kubeadm at least.

Reply:
@munnerz @neolit123 
I'm proposing to apply defaults to the internal types because this IMO provides consistent defaulting across different API versions and makes simpler ussge of both defaults (no need of conversion between public API/internal api) and decoding (yaml->internal->default Vs yaml->public api->default->internal).

Nevertheless, no problem for me to change to a more canonical approach. Just let me known

Reply:
Now that I think to this again there is value in using the "canonical" approach (e.g. avoid confusion in people reading the code).
So I'm moving defaulting back to the versioned config

Reply:
I think the reason to do defaulting in the versioned config is also to opt into versioned behavior, which would be nice to preserve in the future. In the alpha state, we can and will break anything and everything between commits so I'm not too worried about this

Reply:
default is in versioned api now so we are already on the right track